### ISSUE TITLE
The Rakefile fingerprinting is not unique across different Ruby compiles

### ISSUE THREAD

**Issue by 64kramsystem** [tenderlove/tenderjit#36]
In order to know when to (re)generate the Ruby internals data (`lib/tenderjit/ruby`), the Rakefile uses a (shortened) hash of `RUBY_DESCRIPTION`.

When switching between different compilers though (or different compiling options, e.g. adding debug symbols), the internals data changes, while the `RUBY_DESCRIPTION` doesn't.

While switching builds is not very common, there are some use cases (different optimization levels, different compilers...) that it would be useful to cover.

There are different ways of identifying Ruby, I think fingerprinting the binary itself `IO.read(Gem.ruby)` should do the job precisely and simply. If that's a good enough solution, I don't think there's any need to go to great lengths with `File.expand_path`, as that's a far-fetched case (and false positives would just trigger a rebuild).

I can take care of this, as there are only a couple of places to change.

**Comment by tenderlove** [comment]
Probably fingerprinting the binary is fine.  I was mainly worried about startup time.  If fingerprinting the binary itself is fast enough (probably should be with MD5?), then I think it's fine.

**Comment by 64kramsystem** [comment]
Right! I'll check a few other options and time them :+1:

**Comment by tenderlove** [comment]
Another thought, maybe we can make the Rakefile use the mtime for the Ruby binary to know when to "rebuild", but still use the description SHA for determining which file to require.  (I don't really mind what solution we use as long as startup time is not too bad)

**Comment by 64kramsystem** [comment]
I have found what I think is the most effective solution: including, along with the `RUBY_DESCRIPTION`, the values `RbConfig::CONFIG["CC"]` and `RbConfig::CONFIG["CFLAGS"]`. If we really want to be obsessive, there's also `CONFIG["CC_VERSION_MESSAGE"]`, which covers the compiler version (hehe).

I've checked the `rbconfig.rb` source, and the values are placed at Ruby compile time, so there is virtually no speed hit (`RbConfig` is also used somewhere else in TenderJIT, so the time taken by `require` is necessary anyway).  
I've also inspected a prebuilt binary, and `rb_config.rb` was present as well.

I think this covers possibly all the real-world compiler changes. There are certainly obscure edge cases (I'm not familiar with C compiling), but the logic above should cover 128% of the cases :grin:.

On a related note, something I'd consider doing is to extract the fingerprinting version inside a minimal helper module, since the logic is called twice, and it could be considered very simple not trivial anymore, probably into something like `lib/tenderjit/helpers/ruby_version_helper` (`RubyVersionHelper#ruby_fingerprint()`).
